### ISSUE TITLE
Why is prediction error with train_loader and the best model different from train log error ?

### ISSUE THREAD

**Issue by chuzheng88** [Lightning-AI/lightning#13649]
### Environment

- PyTorch Lightning Version (e.g., 1.5.0):
- PyTorch Version (e.g., 1.10):
- Python version (e.g., 3.9):
- OS (e.g., Linux):
- CUDA/cuDNN version:
- GPU models and configuration:
- How you installed PyTorch (`conda`, `pip`, source):
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

My program logs loss/train_loss on every training_epoch_end() function as follows:
![image](https://user-images.githubusercontent.com/17675722/178941844-1409a1c8-c73e-40b2-bbb0-31e8c63c7756.png)

However, logged loss/train_loss calculated on the best model's epoch is different from train_loss calculated by trainer.predict() function, and related function described as follows:
def predict_step(self, batch, batch_idx):
![image](https://user-images.githubusercontent.com/17675722/178942585-1a109737-03ae-40f8-bd33-ae566df237b6.png)
Prediction logic:
![image](https://user-images.githubusercontent.com/17675722/178943050-29c49c2a-4e57-4688-ab4a-4ecc7f1d2d73.png)

I do not why ?

Another information:
logged loss/val_loss calculated on the best model's epoch is same as val_loss calculated by trainer.predict(),. Note these logic same as above and they are described as follows:
![image](https://user-images.githubusercontent.com/17675722/178944234-49e8fb82-ae54-4b17-8d87-d3983fb526ef.png)

![image](https://user-images.githubusercontent.com/17675722/178944142-129fd2c1-511b-45e8-9e3e-3844aeff9da5.png)

**Comment by chuzheng88** [comment]
> > However, logged loss/train_loss calculated on the best model's epoch is different from train_loss calculated by trainer.predict() function,
> 
> 2 reasons:
> 
> 1. because during the epoch, model weights are being updated during training after each step, but during prediction, the weights are consistent.
> 2. Prediction runs on eval mode so that might differ even if the weights are consistent across the epoch.

I have some doubts about this.
I understand that the weights are consistent during prediction, but I log loss/train_loss on def training_epoch_end(self, outputs) function, meaning that the weights have been updated in  my opinon.

**Comment by chuzheng88** [comment]
> > but I log loss/train_loss on def training_epoch_end(self, outputs) function,
> 
> but the values used to compute the final metric are computed inside `training_step`.

@rohitgr7 logged loss/train_loss are computed by mean of all loss inside "training_stemp", 

    def training_epoch_end(self, outputs):
        self.model_train_loss = torch.mean(torch.tensor([o["loss"] for o in outputs]))
        self.log('loss/train_loss', self.model_train_loss)

**Comment by rohitgr7** [comment]
yes, but these `outputs` are computed inside `training_step`. Each `o["loss"]` in `outputs` is computed with a different set of weights.

**Comment by chuzheng88** [comment]
I have another question, can you help me ?

https://github.com/Lightning-AI/lightning/discussions/13648

**Comment by chuzheng88** [comment]
@rohitgr7 I have understand the fact that prediction error with train dataloader and the best model is different from train error calculated by each step. In my case, a regression problem, prediction MSE is 300 with train dataloader and the best model, train MSE calculated by each step is 80.  Is the gap of 220 (300 - 80) normal ? I think that the gap (220 RMSE) is large, but I do not know if this is correct.

**Comment by rohitgr7** [comment]
>  train MSE calculated by each step is 80

are you sure? you can always check the values in outputs since the accumulated final metric is called explicitly by you.

**Comment by chuzheng88** [comment]
> > train MSE calculated by each step is 80
> 
> are you sure? you can always check the values in outputs since the accumulated final metric is called explicitly by you.

Yes, I am sure. On the last epoch, train mse loss is 80.93, ploted as follows:

And then, I calculated the mse with trainloader and the last model is 349, ploted as follows:
![image](https://user-images.githubusercontent.com/17675722/179442426-4516268f-feba-4382-ac00-1361ff9c804f.png)

As said above, I understand that the model's weights are changed every step, meaning that the error with trainloader and the last model different from the error with the same model. However. However,  I think that the gap 270 (350-80) is large, and  I do not know if this is correct.

**Comment by rohitgr7** [comment]
from where did you get the loss info of the second image?

**Comment by chuzheng88** [comment]
> from where did you get the loss info of the second image?

I use the predict_step(self, batch, batch_idx): function to predict result with trainloader and the last model, and then use the results to caluclated MSE （described as the second image）.
main logic:
![image](https://user-images.githubusercontent.com/17675722/179651860-725bad7d-3d8d-4798-a80d-d97171498739.png)

predict_step function:
![image](https://user-images.githubusercontent.com/17675722/179651738-66f2d734-23bb-43be-97e5-62a33b6886d0.png)

_cal_metrics function:
![image](https://user-images.githubusercontent.com/17675722/179651936-0d9b8a43-ebc9-4b37-a075-bc61af235dd6.png)

**Comment by chuzheng88** [comment]
To a certain context, I have solved this problem. In my case, I usede the parameter suffle in train_loader to address this problem.
